src/java.desktop/share/classes/javax/swing/text/html/CSS.java

```diff
@@ -2862,18 +2862,15 @@ ImageIcon getImage(URL base) {
             lengthMapping.put("cm", 28.3464f);
             lengthMapping.put("pc", 12f);
             lengthMapping.put("in", 72f);
-            int res = 72;
-            try {
-                res = Toolkit.getDefaultToolkit().getScreenResolution();
```

Comment:
Are you sure that you can get the exact values of these absolute lengths without using a DPI of the screen? I mean that these values can be measured with a ruler on the screen and they must correspond to "cm" is one centimeter, and "in" is one "inch" etc.

BTW how the browsers handle that?

Reply:
I opened this issue and suggested this fix because it is in line with how `px` were handled already, and fixes convertions between the various units.

> Are you sure that you can get the exact values of these absolute lengths without using a DPI of the screen?

It depends on how Java handles hi-DPI awareness in general, which I am not aware of.
Looking at the sources again, I've found `java.awt.GraphicsConfiguration#getDefaultTransform()` and `java.awt.GraphicsConfiguration#getNormalizingTransform()`, both of which are apparently related to hi-DPI awareness.

I think `getDefaultTransform` is usually used by the Runtime, and `getNormalizingTransform` is what we are looking for, acording to its JavaDoc:
> Returns an AffineTransform that can be concatenated with the default AffineTransform of a GraphicsConfiguration so that 72 units in user space equals 1 inch in device space. [...]

I was quite happy with the results without considering `getNormalizingTransform`, but maybe I was just lucky that it corresponds to the identity transformation for my setup.

> BTW how the browsers handle that?

Browsers have a `devicePixelRatio` indicating how many native device pixels form one CSS pixel (i.e. `px`).
On hi-DPI device, this value is greater than 1, e.g. 2, otherwise it is 1. It depends on the pixel density as well as on the usual viewing distance of the device.
The convertion between the various units is always a constant factor. Units such as `cm` or `in` are never precise, and thus not widely used.